[ICE]: assertion failed: ty::IsRigid::Yes, expected No

<!--
[31mICE[0m: Rustc ./a.rs '-Znext-solver=globally -ooutputfile -Zdump-mir-dir=dir' 'thread 'rustc' ($TID) panicked at compiler/rustc_next_trait_solver/src/solve/assembly/mod.rs:1098:25: 'assertion `left == right` failed'', 'thread 'rustc' ($TID) panicked at compiler/rustc_next_trait_solver/src/solve/assembly/mod.rs:1098:25: 'assertion `left == right` failed''
File: /tmp/im/e/a.rs
-->

auto-reduced (treereduce-rust):
````rust
//@compile-flags: -Znext-solver=globally
type Bar = impl std::fmt::Display;
type MyIter<T> = impl Iterator<Item = T>;
#[define_opaque(MyIter)]
fn my_iter2() -> MyIter<Bar> {}

````

original:
````rust
#![feature(type_alias_impl_trait)]
type Bar = impl std::fmt::Display;
type MyIter<T> = impl Iterator<Item = T>;
#[define_opaque(MyIter)]
fn my_iter2() -> MyIter<Bar> {}

fn main() {}
ty::IsRigid::No
````

Version information
````
rustc 1.98.0-nightly (c397dae80 2026-07-02)
binary: rustc
commit-hash: c397dae808f70caebab1fc4e11b3edf7e59f58c7
commit-date: 2026-07-02
host: x86_64-unknown-linux-gnu
release: 1.98.0-nightly
LLVM version: 22.1.8

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_next_trait_solver/src/solve/assembly/mod.rs#L1092-L1104

Command:
`/home/matthias/.rustup/toolchains/alt-master/bin/rustc -Znext-solver=globally`

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0658]: use of unstable library feature `type_alias_impl_trait`: `type_alias_impl_trait` has open design concerns
 --> /tmp/icemaker_global_tempdir.e8mLYyoP3vmO/rustc_testrunner_tmpdir_reporting.AAn2RmyoKSUG/mvce.rs:3:3
  |
3 | #[define_opaque(MyIter)]
  |   ^^^^^^^^^^^^^
  |
  = note: see issue #63063 <https://github.com/rust-lang/rust/issues/63063> for more information
  = help: add `#![feature(type_alias_impl_trait)]` to the crate attributes to enable
  = note: this compiler was built on 2026-07-02; consider upgrading it if it is out of date

error[E0658]: `impl Trait` in type aliases is unstable
 --> /tmp/icemaker_global_tempdir.e8mLYyoP3vmO/rustc_testrunner_tmpdir_reporting.AAn2RmyoKSUG/mvce.rs:1:12
  |
1 | type Bar = impl std::fmt::Display;
  |            ^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #63063 <https://github.com/rust-lang/rust/issues/63063> for more information
  = help: add `#![feature(type_alias_impl_trait)]` to the crate attributes to enable
  = note: this compiler was built on 2026-07-02; consider upgrading it if it is out of date

error[E0658]: `impl Trait` in type aliases is unstable
 --> /tmp/icemaker_global_tempdir.e8mLYyoP3vmO/rustc_testrunner_tmpdir_reporting.AAn2RmyoKSUG/mvce.rs:2:18
  |
2 | type MyIter<T> = impl Iterator<Item = T>;
  |                  ^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #63063 <https://github.com/rust-lang/rust/issues/63063> for more information
  = help: add `#![feature(type_alias_impl_trait)]` to the crate attributes to enable
  = note: this compiler was built on 2026-07-02; consider upgrading it if it is out of date

error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.e8mLYyoP3vmO/rustc_testrunner_tmpdir_reporting.AAn2RmyoKSUG/mvce.rs:4:32
  |
4 | fn my_iter2() -> MyIter<Bar> {}
  |                                ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.e8mLYyoP3vmO/rustc_testrunner_tmpdir_reporting.AAn2RmyoKSUG/mvce.rs`

error: unconstrained opaque type
 --> /tmp/icemaker_global_tempdir.e8mLYyoP3vmO/rustc_testrunner_tmpdir_reporting.AAn2RmyoKSUG/mvce.rs:1:12
  |
1 | type Bar = impl std::fmt::Display;
  |            ^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `Bar` must be used in combination with a concrete type within the same crate


thread 'rustc' (1931447) panicked at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_next_trait_solver/src/solve/assembly/mod.rs:1098:25:
assertion `left == right` failed
  left: Yes
 right: No
stack backtrace:
   0:     0x7f34f4015060 - std[5174af7d8d584da7]::backtrace_rs::backtrace::libunwind::trace
                               at /rustc/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/std/src/../../backtrace/src/backtrace/libunwind.rs:117:9
   1:     0x7f34f4015060 - std[5174af7d8d584da7]::backtrace_rs::backtrace::trace_unsynchronized::<std[5174af7d8d584da7]::sys::backtrace::_print_fmt::{closure#1}>
                               at /rustc/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/std/src/../../backtrace/src/backtrace/mod.rs:66:14
   2:     0x7f34f4015060 - std[5174af7d8d584da7]::sys::backtrace::_print_fmt
                               at /rustc/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/std/src/sys/backtrace.rs:74:9
   3:     0x7f34f4015060 - <<std[5174af7d8d584da7]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[9cd6bca4bf086ba5]::fmt::Display>::fmt
                               at /rustc/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/std/src/sys/backtrace.rs:44:26
   4:     0x7f34ef890aee - <core[9cd6bca4bf086ba5]::fmt::rt::Argument>::fmt
                               at /rustc/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/core/src/fmt/rt.rs:152:76
   5:     0x7f34ef890aee - core[9cd6bca4bf086ba5]::fmt::write
   6:     0x7f34f402c74c - std[5174af7d8d584da7]::io::default_write_fmt::<std[5174af7d8d584da7]::sys::stdio::unix::Stderr>
                               at /rustc/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/std/src/io/mod.rs:626:11
   7:     0x7f34f402c74c - <std[5174af7d8d584da7]::sys::stdio::unix::Stderr as std[5174af7d8d584da7]::io::Write>::write_fmt
                               at /rustc/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/std/src/io/mod.rs:1730:13
   8:     0x7f34f3fd5ee4 - <std[5174af7d8d584da7]::sys::backtrace::BacktraceLock>::print
                               at /rustc/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/std/src/sys/backtrace.rs:47:9
   9:     0x7f34f3fd5ee4 - std[5174af7d8d584da7]::panicking::default_hook::{closure#0}
                               at /rustc/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/std/src/panicking.rs:292:27
  10:     0x7f34f4001661 - std[5174af7d8d584da7]::panicking::default_hook
                               at /rustc/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/std/src/panicking.rs:319:9
  11:     0x7f34f0bb7d18 - <alloc[ba8af95fd7c02708]::boxed::Box<dyn for<'a, 'b> core[9cd6bca4bf086ba5]::ops::function::Fn<(&'a std[5174af7d8d584da7]::panic::PanicHookInfo<'b>,), Output = ()> + core[9cd6bca4bf086ba5]::marker::Send + core[9cd6bca4bf086ba5]::marker::Sync> as core[9cd6bca4bf086ba5]::ops::function::Fn<(&std[5174af7d8d584da7]::panic::PanicHookInfo,)>>::call
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/alloc/src/boxed.rs:2333:9
  12:     0x7f34f0bb7d18 - rustc_driver_impl[684992a3f0f3f303]::install_ice_hook::{closure#1}
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_driver_impl/src/lib.rs:1483:17
  13:     0x7f34f0bb7d18 - <alloc[ba8af95fd7c02708]::boxed::Box<rustc_driver_impl[684992a3f0f3f303]::install_ice_hook::{closure#1}> as core[9cd6bca4bf086ba5]::ops::function::Fn<(&dyn for<'a, 'b> core[9cd6bca4bf086ba5]::ops::function::Fn<(&'a std[5174af7d8d584da7]::panic::PanicHookInfo<'b>,), Output = ()> + core[9cd6bca4bf086ba5]::marker::Send + core[9cd6bca4bf086ba5]::marker::Sync, &std[5174af7d8d584da7]::panic::PanicHookInfo)>>::call
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/alloc/src/boxed.rs:2333:9
  14:     0x7f34f0bb7d18 - std[5174af7d8d584da7]::panicking::update_hook::<alloc[ba8af95fd7c02708]::boxed::Box<rustc_driver_impl[684992a3f0f3f303]::install_ice_hook::{closure#1}>>::{closure#0}
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/std/src/panicking.rs:235:47
  15:     0x7f34f4001aea - <alloc[ba8af95fd7c02708]::boxed::Box<dyn for<'a, 'b> core[9cd6bca4bf086ba5]::ops::function::Fn<(&'a std[5174af7d8d584da7]::panic::PanicHookInfo<'b>,), Output = ()> + core[9cd6bca4bf086ba5]::marker::Send + core[9cd6bca4bf086ba5]::marker::Sync> as core[9cd6bca4bf086ba5]::ops::function::Fn<(&std[5174af7d8d584da7]::panic::PanicHookInfo,)>>::call
                               at /rustc/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/alloc/src/boxed.rs:2333:9
  16:     0x7f34f4001aea - std[5174af7d8d584da7]::panicking::panic_with_hook
                               at /rustc/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/std/src/panicking.rs:823:13
  17:     0x7f34f3fd5fb2 - std[5174af7d8d584da7]::panicking::panic_handler::{closure#0}
                               at /rustc/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/std/src/panicking.rs:688:13
  18:     0x7f34f3fc14a9 - std[5174af7d8d584da7]::sys::backtrace::__rust_end_short_backtrace::<std[5174af7d8d584da7]::panicking::panic_handler::{closure#0}, !>
                               at /rustc/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/std/src/sys/backtrace.rs:182:18
  19:     0x7f34f3fd7ebd - __rustc[41205345b0684f81]::rust_begin_unwind
                               at /rustc/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/std/src/panicking.rs:679:5
  20:     0x7f34ef89181c - core[9cd6bca4bf086ba5]::panicking::panic_fmt
                               at /rustc/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/core/src/panicking.rs:80:14
  21:     0x7f34ef8916a3 - core[9cd6bca4bf086ba5]::panicking::assert_failed_inner
                               at /rustc/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/core/src/panicking.rs:439:17
  22:     0x7f34f17c4b13 - core[9cd6bca4bf086ba5]::panicking::assert_failed::<rustc_type_ir[edb981713e479e19]::ty_kind::IsRigid, rustc_type_ir[edb981713e479e19]::ty_kind::IsRigid>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/core/src/panicking.rs:394:5
  23:     0x7f34f3be1ab3 - <<rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<_, _>>::try_assemble_bounds_via_registered_opaques::ReplaceOpaque<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt> as rustc_type_ir[edb981713e479e19]::fold::TypeFolder<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::fold_ty
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_next_trait_solver/src/solve/assembly/mod.rs:1098:25
  24:     0x7f34f3982450 - <rustc_middle[38b942353e9406ad]::ty::Ty as rustc_type_ir[edb981713e479e19]::fold::TypeFoldable<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::fold_with::<<rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<_, _>>::try_assemble_bounds_via_registered_opaques::ReplaceOpaque<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_middle/src/ty/structural_impls.rs:337:16
  25:     0x7f34f3982450 - <rustc_middle[38b942353e9406ad]::ty::Term as rustc_type_ir[edb981713e479e19]::fold::TypeFoldable<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::fold_with::<<rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<_, _>>::try_assemble_bounds_via_registered_opaques::ReplaceOpaque<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_middle/src/ty/mod.rs:617:40
  26:     0x7f34f3982450 - <rustc_type_ir[edb981713e479e19]::predicate::ProjectionPredicate<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt> as rustc_type_ir[edb981713e479e19]::fold::TypeFoldable<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::fold_with::<<rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<_, _>>::try_assemble_bounds_via_registered_opaques::ReplaceOpaque<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_type_ir/src/predicate.rs:552:55
  27:     0x7f34f3982450 - <rustc_type_ir[edb981713e479e19]::predicate_kind::ClauseKind<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt> as rustc_type_ir[edb981713e479e19]::fold::TypeFoldable<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::fold_with::<<rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<_, _>>::try_assemble_bounds_via_registered_opaques::ReplaceOpaque<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_type_ir/src/predicate_kind.rs:13:55
  28:     0x7f34f3982450 - <rustc_type_ir[edb981713e479e19]::predicate_kind::PredicateKind<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt> as rustc_type_ir[edb981713e479e19]::fold::TypeFoldable<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::fold_with::<<rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<_, _>>::try_assemble_bounds_via_registered_opaques::ReplaceOpaque<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_type_ir/src/predicate_kind.rs:61:55
  29:     0x7f34f3982450 - <rustc_type_ir[edb981713e479e19]::binder::Binder<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt, rustc_type_ir[edb981713e479e19]::predicate_kind::PredicateKind<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>> as rustc_type_ir[edb981713e479e19]::fold::TypeSuperFoldable<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::super_fold_with::<<rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<_, _>>::try_assemble_bounds_via_registered_opaques::ReplaceOpaque<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::{closure#0}
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_type_ir/src/binder.rs:128:30
  30:     0x7f34f3982450 - <rustc_type_ir[edb981713e479e19]::binder::Binder<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt, rustc_type_ir[edb981713e479e19]::predicate_kind::PredicateKind<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>>::map_bound::<<rustc_type_ir[edb981713e479e19]::binder::Binder<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt, rustc_type_ir[edb981713e479e19]::predicate_kind::PredicateKind<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>> as rustc_type_ir[edb981713e479e19]::fold::TypeSuperFoldable<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::super_fold_with<<rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<_, _>>::try_assemble_bounds_via_registered_opaques::ReplaceOpaque<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::{closure#0}, rustc_type_ir[edb981713e479e19]::predicate_kind::PredicateKind<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_type_ir/src/binder.rs:183:21
  31:     0x7f34f3982450 - <rustc_type_ir[edb981713e479e19]::binder::Binder<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt, rustc_type_ir[edb981713e479e19]::predicate_kind::PredicateKind<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>> as rustc_type_ir[edb981713e479e19]::fold::TypeSuperFoldable<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::super_fold_with::<<rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<_, _>>::try_assemble_bounds_via_registered_opaques::ReplaceOpaque<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_type_ir/src/binder.rs:128:14
  32:     0x7f34f3982450 - <<rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<_, _>>::try_assemble_bounds_via_registered_opaques::ReplaceOpaque<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt> as rustc_type_ir[edb981713e479e19]::fold::TypeFolder<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::fold_binder::<rustc_type_ir[edb981713e479e19]::predicate_kind::PredicateKind<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_type_ir/src/fold.rs:131:11
  33:     0x7f34f3982450 - <rustc_type_ir[edb981713e479e19]::binder::Binder<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt, rustc_type_ir[edb981713e479e19]::predicate_kind::PredicateKind<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>> as rustc_type_ir[edb981713e479e19]::fold::TypeFoldable<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::fold_with::<<rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<_, _>>::try_assemble_bounds_via_registered_opaques::ReplaceOpaque<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_type_ir/src/binder.rs:109:16
  34:     0x7f34f3982450 - <rustc_middle[38b942353e9406ad]::ty::predicate::Predicate as rustc_type_ir[edb981713e479e19]::fold::TypeSuperFoldable<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::super_fold_with::<<rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<_, _>>::try_assemble_bounds_via_registered_opaques::ReplaceOpaque<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_middle/src/ty/structural_impls.rs:575:31
  35:     0x7f34f3982450 - <<rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<_, _>>::try_assemble_bounds_via_registered_opaques::ReplaceOpaque<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt> as rustc_type_ir[edb981713e479e19]::fold::TypeFolder<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::fold_predicate
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_type_ir/src/fold.rs:149:11
  36:     0x7f34f3982450 - <rustc_middle[38b942353e9406ad]::ty::predicate::Clause as rustc_type_ir[edb981713e479e19]::fold::TypeFoldable<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::fold_with::<<rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<_, _>>::try_assemble_bounds_via_registered_opaques::ReplaceOpaque<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_middle/src/ty/structural_impls.rs:529:16
  37:     0x7f34f3a79eb3 - <rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[8e9523eaf1ac6a96]::solve::delegate::SolverDelegate, rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::try_assemble_bounds_via_registered_opaques::<rustc_type_ir[edb981713e479e19]::predicate::TraitPredicate<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_next_trait_solver/src/solve/assembly/mod.rs:1121:32
  38:     0x7f34f3a79eb3 - <rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[8e9523eaf1ac6a96]::solve::delegate::SolverDelegate, rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::assemble_and_evaluate_candidates::<rustc_type_ir[edb981713e479e19]::predicate::TraitPredicate<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_next_trait_solver/src/solve/assembly/mod.rs:450:18
  39:     0x7f34f3a79eb3 - <rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[8e9523eaf1ac6a96]::solve::delegate::SolverDelegate, rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::compute_trait_goal
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_next_trait_solver/src/solve/trait_goals.rs:1576:18
  40:     0x7f34f39d154b - <rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[8e9523eaf1ac6a96]::solve::delegate::SolverDelegate, rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::compute_goal::{closure#0}
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_next_trait_solver/src/solve/eval_ctxt/mod.rs:927:25
  41:     0x7f34f39eb1d4 - <rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[8e9523eaf1ac6a96]::solve::delegate::SolverDelegate, rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::enter_forall_with_assumptions::<rustc_type_ir[edb981713e479e19]::predicate_kind::PredicateKind<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>, core[9cd6bca4bf086ba5]::result::Result<rustc_type_ir[edb981713e479e19]::canonical::Canonical<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt, rustc_type_ir[edb981713e479e19]::solve::Response<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>, rustc_type_ir[edb981713e479e19]::solve::NoSolutionOrRerunNonErased>, <rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[8e9523eaf1ac6a96]::solve::delegate::SolverDelegate, rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::compute_goal::{closure#0}>::{closure#0}
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_next_trait_solver/src/solve/eval_ctxt/mod.rs:1334:13
  42:     0x7f34f39eb1d4 - <rustc_infer[bd31e28fd3790da]::infer::InferCtxt>::enter_forall::<rustc_type_ir[edb981713e479e19]::predicate_kind::PredicateKind<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>, core[9cd6bca4bf086ba5]::result::Result<rustc_type_ir[edb981713e479e19]::canonical::Canonical<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt, rustc_type_ir[edb981713e479e19]::solve::Response<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>, rustc_type_ir[edb981713e479e19]::solve::NoSolutionOrRerunNonErased>, <rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[8e9523eaf1ac6a96]::solve::delegate::SolverDelegate, rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::enter_forall_with_assumptions<rustc_type_ir[edb981713e479e19]::predicate_kind::PredicateKind<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>, core[9cd6bca4bf086ba5]::result::Result<rustc_type_ir[edb981713e479e19]::canonical::Canonical<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt, rustc_type_ir[edb981713e479e19]::solve::Response<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>, rustc_type_ir[edb981713e479e19]::solve::NoSolutionOrRerunNonErased>, <rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[8e9523eaf1ac6a96]::solve::delegate::SolverDelegate, rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::compute_goal::{closure#0}>::{closure#0}>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_infer/src/infer/relate/higher_ranked.rs:76:9
  43:     0x7f34f39eb1d4 - <rustc_infer[bd31e28fd3790da]::infer::InferCtxt as rustc_type_ir[edb981713e479e19]::infer_ctxt::InferCtxtLike>::enter_forall_without_assumptions::<rustc_type_ir[edb981713e479e19]::predicate_kind::PredicateKind<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>, core[9cd6bca4bf086ba5]::result::Result<rustc_type_ir[edb981713e479e19]::canonical::Canonical<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt, rustc_type_ir[edb981713e479e19]::solve::Response<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>, rustc_type_ir[edb981713e479e19]::solve::NoSolutionOrRerunNonErased>, <rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[8e9523eaf1ac6a96]::solve::delegate::SolverDelegate, rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::enter_forall_with_assumptions<rustc_type_ir[edb981713e479e19]::predicate_kind::PredicateKind<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>, core[9cd6bca4bf086ba5]::result::Result<rustc_type_ir[edb981713e479e19]::canonical::Canonical<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt, rustc_type_ir[edb981713e479e19]::solve::Response<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>, rustc_type_ir[edb981713e479e19]::solve::NoSolutionOrRerunNonErased>, <rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[8e9523eaf1ac6a96]::solve::delegate::SolverDelegate, rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::compute_goal::{closure#0}>::{closure#0}>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_infer/src/infer/context.rs:217:14
  44:     0x7f34f39eb1d4 - <rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[8e9523eaf1ac6a96]::solve::delegate::SolverDelegate, rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::enter_forall_with_assumptions::<rustc_type_ir[edb981713e479e19]::predicate_kind::PredicateKind<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>, core[9cd6bca4bf086ba5]::result::Result<rustc_type_ir[edb981713e479e19]::canonical::Canonical<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt, rustc_type_ir[edb981713e479e19]::solve::Response<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>, rustc_type_ir[edb981713e479e19]::solve::NoSolutionOrRerunNonErased>, <rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[8e9523eaf1ac6a96]::solve::delegate::SolverDelegate, rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::compute_goal::{closure#0}>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_next_trait_solver/src/solve/eval_ctxt/mod.rs:1326:23
  45:     0x7f34f39eb1d4 - <rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[8e9523eaf1ac6a96]::solve::delegate::SolverDelegate, rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::compute_goal
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_next_trait_solver/src/solve/eval_ctxt/mod.rs:924:14
  46:     0x7f34f39eb1d4 - <rustc_next_trait_solver[e298119d98126207]::solve::search_graph::SearchGraphDelegate<rustc_trait_selection[8e9523eaf1ac6a96]::solve::delegate::SolverDelegate> as rustc_type_ir[edb981713e479e19]::search_graph::Delegate>::compute_goal::{closure#0}::{closure#0}
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_next_trait_solver/src/solve/search_graph.rs:145:34
  47:     0x7f34f39eb1d4 - <rustc_next_trait_solver[e298119d98126207]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[8e9523eaf1ac6a96]::solve::delegate::SolverDelegate, rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::enter_canonical::<rustc_type_ir[edb981713e479e19]::canonical::Canonical<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt, rustc_type_ir[edb981713e479e19]::solve::Response<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>, <rustc_next_trait_solver[e298119d98126207]::solve::search_graph::SearchGraphDelegate<rustc_trait_selection[8e9523eaf1ac6a96]::solve::delegate::SolverDelegate> as rustc_type_ir[edb981713e479e19]::search_graph::Delegate>::compute_goal::{closure#0}::{closure#0}>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_next_trait_solver/src/solve/eval_ctxt/mod.rs:444:22
  48:     0x7f34f39eb1d4 - <rustc_next_trait_solver[e298119d98126207]::solve::search_graph::SearchGraphDelegate<rustc_trait_selection[8e9523eaf1ac6a96]::solve::delegate::SolverDelegate> as rustc_type_ir[edb981713e479e19]::search_graph::Delegate>::compute_goal::{closure#0}
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_next_trait_solver/src/solve/search_graph.rs:143:13
  49:     0x7f34f3adc737 - stacker[3c3bcd11e6ec8e9c]::maybe_grow::<(core[9cd6bca4bf086ba5]::result::Result<rustc_type_ir[edb981713e479e19]::canonical::Canonical<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt, rustc_type_ir[edb981713e479e19]::solve::Response<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>, rustc_type_ir[edb981713e479e19]::solve::NoSolution>, rustc_type_ir[edb981713e479e19]::solve::AccessedOpaques<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>), <rustc_next_trait_solver[e298119d98126207]::solve::search_graph::SearchGraphDelegate<rustc_trait_selection[8e9523eaf1ac6a96]::solve::delegate::SolverDelegate> as rustc_type_ir[edb981713e479e19]::search_graph::Delegate>::compute_goal::{closure#0}>
                               at /rust/deps/stacker-0.1.21/src/lib.rs:57:9
  50:     0x7f34f3adc737 - rustc_data_structures[19b5b55f2b7477ab]::stack::ensure_sufficient_stack::<(core[9cd6bca4bf086ba5]::result::Result<rustc_type_ir[edb981713e479e19]::canonical::Canonical<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt, rustc_type_ir[edb981713e479e19]::solve::Response<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>, rustc_type_ir[edb981713e479e19]::solve::NoSolution>, rustc_type_ir[edb981713e479e19]::solve::AccessedOpaques<rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>), <rustc_next_trait_solver[e298119d98126207]::solve::search_graph::SearchGraphDelegate<rustc_trait_selection[8e9523eaf1ac6a96]::solve::delegate::SolverDelegate> as rustc_type_ir[edb981713e479e19]::search_graph::Delegate>::compute_goal::{closure#0}>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_data_structures/src/stack.rs:21:5
  51:     0x7f34f3adc737 - <rustc_next_trait_solver[e298119d98126207]::solve::search_graph::SearchGraphDelegate<rustc_trait_selection[8e9523eaf1ac6a96]::solve::delegate::SolverDelegate> as rustc_type_ir[edb981713e479e19]::search_graph::Delegate>::compute_goal
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_next_trait_solver/src/solve/search_graph.rs:142:9
  52:     0x7f34f3adc737 - <rustc_type_ir[edb981713e479e19]::search_graph::SearchGraph<rustc_next_trait_solver[e298119d98126207]::solve::search_graph::SearchGraphDelegate<rustc_trait_selection[8e9523eaf1ac6a96]::solve::delegate::SolverDelegate>, rustc_middle[38b942353e9406ad]::ty::context::TyCtxt>>::evaluate_goal_in_task
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_type_ir/src/search_graph/mod.rs:1349:26

 ...
 
 446:     0x7f34f0bab295 - rustc_span[7b02b36b5f73caf4]::create_session_globals_then::<(), rustc_interface[c472725c2f360e16]::util::run_in_thread_with_globals<rustc_interface[c472725c2f360e16]::util::run_in_thread_pool_with_globals<rustc_interface[c472725c2f360e16]::interface::run_compiler<(), rustc_driver_impl[684992a3f0f3f303]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_span/src/lib.rs:155:21
 447:     0x7f34f0bab295 - rustc_interface[c472725c2f360e16]::util::run_in_thread_with_globals::<rustc_interface[c472725c2f360e16]::util::run_in_thread_pool_with_globals<rustc_interface[c472725c2f360e16]::interface::run_compiler<(), rustc_driver_impl[684992a3f0f3f303]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/compiler/rustc_interface/src/util.rs:158:17
 448:     0x7f34f0bab295 - std[5174af7d8d584da7]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[c472725c2f360e16]::util::run_in_thread_with_globals<rustc_interface[c472725c2f360e16]::util::run_in_thread_pool_with_globals<rustc_interface[c472725c2f360e16]::interface::run_compiler<(), rustc_driver_impl[684992a3f0f3f303]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/std/src/sys/backtrace.rs:166:18
 449:     0x7f34f0bbfdf9 - std[5174af7d8d584da7]::thread::lifecycle::spawn_unchecked::<rustc_interface[c472725c2f360e16]::util::run_in_thread_with_globals<rustc_interface[c472725c2f360e16]::util::run_in_thread_pool_with_globals<rustc_interface[c472725c2f360e16]::interface::run_compiler<(), rustc_driver_impl[684992a3f0f3f303]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1}::{closure#0}
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/std/src/thread/lifecycle.rs:70:13
 450:     0x7f34f0bbfdf9 - <core[9cd6bca4bf086ba5]::panic::unwind_safe::AssertUnwindSafe<std[5174af7d8d584da7]::thread::lifecycle::spawn_unchecked<rustc_interface[c472725c2f360e16]::util::run_in_thread_with_globals<rustc_interface[c472725c2f360e16]::util::run_in_thread_pool_with_globals<rustc_interface[c472725c2f360e16]::interface::run_compiler<(), rustc_driver_impl[684992a3f0f3f303]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1}::{closure#0}> as core[9cd6bca4bf086ba5]::ops::function::FnOnce<()>>::call_once
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/core/src/panic/unwind_safe.rs:275:9
 451:     0x7f34f0bbfdf9 - std[5174af7d8d584da7]::panicking::catch_unwind::do_call::<core[9cd6bca4bf086ba5]::panic::unwind_safe::AssertUnwindSafe<std[5174af7d8d584da7]::thread::lifecycle::spawn_unchecked<rustc_interface[c472725c2f360e16]::util::run_in_thread_with_globals<rustc_interface[c472725c2f360e16]::util::run_in_thread_pool_with_globals<rustc_interface[c472725c2f360e16]::interface::run_compiler<(), rustc_driver_impl[684992a3f0f3f303]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1}::{closure#0}>, ()>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/std/src/panicking.rs:576:43
 452:     0x7f34f0bbfdf9 - std[5174af7d8d584da7]::panicking::catch_unwind::<(), core[9cd6bca4bf086ba5]::panic::unwind_safe::AssertUnwindSafe<std[5174af7d8d584da7]::thread::lifecycle::spawn_unchecked<rustc_interface[c472725c2f360e16]::util::run_in_thread_with_globals<rustc_interface[c472725c2f360e16]::util::run_in_thread_pool_with_globals<rustc_interface[c472725c2f360e16]::interface::run_compiler<(), rustc_driver_impl[684992a3f0f3f303]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1}::{closure#0}>>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/std/src/panicking.rs:544:19
 453:     0x7f34f0bbfdf9 - std[5174af7d8d584da7]::panic::catch_unwind::<core[9cd6bca4bf086ba5]::panic::unwind_safe::AssertUnwindSafe<std[5174af7d8d584da7]::thread::lifecycle::spawn_unchecked<rustc_interface[c472725c2f360e16]::util::run_in_thread_with_globals<rustc_interface[c472725c2f360e16]::util::run_in_thread_pool_with_globals<rustc_interface[c472725c2f360e16]::interface::run_compiler<(), rustc_driver_impl[684992a3f0f3f303]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1}::{closure#0}>, ()>
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/std/src/panic.rs:359:14
 454:     0x7f34f0bbfdf9 - std[5174af7d8d584da7]::thread::lifecycle::spawn_unchecked::<rustc_interface[c472725c2f360e16]::util::run_in_thread_with_globals<rustc_interface[c472725c2f360e16]::util::run_in_thread_pool_with_globals<rustc_interface[c472725c2f360e16]::interface::run_compiler<(), rustc_driver_impl[684992a3f0f3f303]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1}
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/std/src/thread/lifecycle.rs:68:26
 455:     0x7f34f0bbfdf9 - <std[5174af7d8d584da7]::thread::lifecycle::spawn_unchecked<rustc_interface[c472725c2f360e16]::util::run_in_thread_with_globals<rustc_interface[c472725c2f360e16]::util::run_in_thread_pool_with_globals<rustc_interface[c472725c2f360e16]::interface::run_compiler<(), rustc_driver_impl[684992a3f0f3f303]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[9cd6bca4bf086ba5]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
                               at /rustc-dev/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/core/src/ops/function.rs:250:5
 456:     0x7f34f4010b7f - <alloc[ba8af95fd7c02708]::boxed::Box<dyn core[9cd6bca4bf086ba5]::ops::function::FnOnce<(), Output = ()> + core[9cd6bca4bf086ba5]::marker::Send> as core[9cd6bca4bf086ba5]::ops::function::FnOnce<()>>::call_once
                               at /rustc/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/alloc/src/boxed.rs:2319:9
 457:     0x7f34f4010b7f - <std[5174af7d8d584da7]::sys::thread::unix::Thread>::new::thread_start
                               at /rustc/c397dae808f70caebab1fc4e11b3edf7e59f58c7/library/std/src/sys/thread/unix.rs:123:17
 458:     0x7f34ede981b9 - <unknown>
 459:     0x7f34edf1d21c - <unknown>
 460:                0x0 - <unknown>

error: the compiler unexpectedly panicked. This is a bug

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.98.0-nightly (c397dae80 2026-07-02) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z next-solver=globally -Z dump-mir-dir=dir

query stack during panic:
#0 [typeck_root] type-checking `my_iter2`
#1 [check_match] match-checking `my_iter2`
#2 [mir_built] building MIR for `my_iter2`
#3 [has_ffi_unwind_calls] checking if `my_iter2` contains FFI-unwind calls
#4 [mir_promoted] promoting constants in MIR for `my_iter2`
#5 [mir_borrowck] borrow-checking `my_iter2`
#6 [type_of_opaque] computing type of opaque `MyIter::{opaque#0}`
#7 [type_of] computing type of `MyIter::{opaque#0}`
#8 [check_well_formed] checking that `MyIter::{opaque#0}` is well-formed
#9 [check_type_wf] checking that types are well-formed
#10 [analysis] running analysis passes on crate `mvce`
end of query stack
error: aborting due to 5 previous errors

Some errors have detailed explanations: E0601, E0658.
For more information about an error, try `rustc --explain E0601`.

```

</p>
</details>


@rustbot label +F-type_alias_impl_trait +WG-trait-system-refactor

--- a/compiler/rustc_next_trait_solver/src/solve/assembly/mod.rs
+++ b/compiler/rustc_next_trait_solver/src/solve/assembly/mod.rs
@@ -1095,8 +1095,8 @@ where
                     if let ty::Alias(is_rigid, alias_ty) = ty.kind()
                         && let Some(opaque_ty) = alias_ty.try_to_opaque()
                     {
-                        debug_assert_eq!(is_rigid, ty::IsRigid::No);
                         if opaque_ty == self.opaque_ty {
+                            debug_assert_eq!(is_rigid, ty::IsRigid::No);
                             return self.self_ty;
                         }
                     }

--- a/tests/ui/traits/next-solver/opaques/dont-ice-replacing-non-rigid-opaque.rs
+++ b/tests/ui/traits/next-solver/opaques/dont-ice-replacing-non-rigid-opaque.rs
@@ -1,0 +1,25 @@
+//@ compile-flags: -Znext-solver
+//@ check-pass
+
+// Regression test for #158784
+//
+// We should assert that only opaque types that is to be replaced
+// are non-rigid. We can have rigid opaque types elsewhere.
+
+#![feature(type_alias_impl_trait)]
+type Rigid = impl Sized;
+#[define_opaque(Rigid)]
+fn define_rigid() -> Rigid {}
+
+type MyIter<T> = impl Iterator<Item = T>;
+
+#[define_opaque(MyIter)]
+fn define_my_iter<T>(a: T) -> MyIter<T> {
+    if false {
+        // `Rigid` being rigid is totally fine here.
+        let _: MyIter<Rigid> = std::iter::once(define_rigid());
+    }
+    std::iter::once(a)
+}
+
+fn main() {}

